api/oc_etag.c

```diff
@@ -0,0 +1,324 @@
+/****************************************************************************
+ *
+ * Copyright (c) 2023 plgd.dev s.r.o.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"),
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ ***************************************************************************/
+
+#include "util/oc_features.h"
+
+#ifdef OC_HAS_FEATURE_ETAG
+
+#include "api/oc_etag_internal.h"
+#include "api/oc_resource_internal.h"
+#include "oc_core_res.h"
+#include "oc_rep.h"
+#include "oc_ri.h"
+#include "port/oc_clock.h"
+#include "port/oc_log_internal.h"
+#include "port/oc_random.h"
+#include "util/oc_macros_internal.h"
+
+#ifdef OC_STORAGE
+#include "api/oc_storage_internal.h"
+
+#ifdef OC_COLLECTIONS
+#include "api/oc_collection_internal.h"
+#endif /* OC_COLLECTIONS */
+
+#endif /* OC_STORAGE */
+
+#include <inttypes.h>
+#include <stdint.h>
+
+static uint64_t g_etag = 0;
+
+static uint64_t
+etag_random(void)
+{
+  return (oc_random_value() % 1000) + 1;
+}
+
+uint64_t
+oc_etag_global(void)
+{
+  return g_etag;
+}
+
+uint64_t
+oc_etag_get(void)
+{
+  uint64_t now = oc_clock_time();
+  if (now > g_etag) {
+    g_etag = now;
+  }
+  g_etag += etag_random();
+  if (g_etag == OC_ETAG_UNINITALIZED) {
+    g_etag = 1;
+  }
+  return g_etag;
```

Comment:
just add here: `TODO: check for overflow` - all resources need to reinitialize etags.